openmls/src/group/mls_group/validation.rs

```diff
@@ -0,0 +1,244 @@
+//! This module contains validation functions for incoming messages
+//! as defined in https://github.com/openmls/openmls/wiki/Message-validation
+
+use std::collections::HashSet;
+
+use super::{proposals::StagedProposalQueue, *};
+
+impl MlsGroup {
+    // === Messages ===
+
+    /// Checks the following semantic validation:
+    ///  - ValSem2
+    ///  - ValSem3
+    pub fn validate_framing(&self, message: &MlsMessageIn) -> Result<(), MlsGroupError> {
+        // ValSem2
+        if message.group_id() != self.group_id() {
+            return Err(FramingValidationError::WrongGroupId.into());
+        }
+
+        // ValSem3
+        if message.epoch() != self.context().epoch() {
+            return Err(FramingValidationError::WrongEpoch.into());
+        }
+
+        Ok(())
+    }
+
+    /// Checks the following semantic validation:
+    ///  - ValSem4
+    ///  - ValSem5
+    ///  - ValSem7
+    ///  - ValSem9
+    pub fn validate_plaintext(
+        &self,
+        plaintext: &VerifiableMlsPlaintext,
+    ) -> Result<(), MlsGroupError> {
+        // ValSem4
+        let sender = plaintext.sender();
+        if sender.is_member() {
+            let sender_index = sender.to_leaf_index();
+            if sender_index > self.tree().leaf_count() || self.tree().nodes[sender_index].is_blank()
+            {
+                return Err(FramingValidationError::UnknownMember.into());
+            }
+        }
+
+        // ValSem5
+        if plaintext.content_type() == ContentType::Application {
+            if plaintext.wire_format() != WireFormat::MlsCiphertext {
+                return Err(FramingValidationError::UnencryptedApplicationMessage.into());
+            } else if !plaintext.sender().is_member() {
+                return Err(FramingValidationError::NonMemberApplicationMessage.into());
+            }
+        }
+
+        // ValSem7
+        if plaintext.sender().is_member()
```

Comment:
The label just says "Membership tag presence". Maybe it would be worthwhile adding the conditions under which a membership tag actually has to be present? This seems to be done for other labels. Same for the confirmation tag presence test below.

Reply:
I added some more comments.

Reply:
I meant changing the actual label attached to ValSem7. Since the current official index of checks with their associated labels is the PR description, my suggestion was to change it there. I.e. "Membership tag presence" -> "Membership tag presence for plaintext messages from group members". Generally, the style of label is not uniform, but maybe we should deal with the labels when we commit them to the rust book or similar.